File: injector/src/test/java/net/mready/inject/PolymorphicDependencyTest.java
package net.mready.inject;

import org.junit.Test;

import javax.inject.Inject;
import javax.inject.Named;

import static org.junit.Assert.assertEquals;

public class PolymorphicDependencyTest {
    @Test
    public void multipleImplementations() {
        Injector injector = Injector.with(new Module());
        assertEquals(FooA.class, injector.instance(Key.of(Foo.class, "A")).getClass());
        assertEquals(FooB.class, injector.instance(Key.of(Foo.class, "B")).getClass());
    }

    public static class Module {
        @Provides
        @Named("A")
        Foo a(FooA fooA) {
            return fooA;
        }

        @Provides
        @Named("B")
        Foo a(FooB fooB) {
            return fooB;
        }
    }

    interface Foo {

    }

    public static class FooA implements Foo {
        @Inject
        public FooA() {
        }
    }

    public static class FooB implements Foo {
        @Inject
        public FooB() {
        }

    }
}

File: injector/src/test/kotlin/net/mready/inject/PolymorphicDependencyTest.kt
package net.mready.inject

import org.junit.Assert.assertEquals
import org.junit.Test
import javax.inject.Named

class PolymorphicDependencyTest {
    @Test
    fun multipleImplementations() {
        val injector = Injector.with(Module())
        assertEquals(FooA::class.java, injector.instance(Key.of(Foo::class.java, "A")).javaClass)
        assertEquals(FooB::class.java, injector.instance(Key.of(Foo::class.java, "B")).javaClass)
    }

    class Module {
        @Provides
        @Named("A") fun a(fooA: FooA): Foo {
            return fooA
        }

        @Provides
        @Named("B") fun a(fooB: FooB): Foo {
            return fooB
        }
    }

    interface Foo

    class FooA : Foo

    class FooB : Foo
}
